rfcs/0128-selective-auto-merge-on-bot-upgrades.md

```diff
@@ -0,0 +1,56 @@
+---
+feature: Selective auto-merge on bot upgrades
+start-date: 2022-07-07
+author: superherointj
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Introduce a `meta.autoMerge` attribute to packages to allow committers to delegate merge rights to package maintainers for bot auto updates.
+
+# Motivation
+[motivation]: #motivation
+
+* Reduce pending PRs for review and merge in nixpkgs.
+* Save reviewers and commiters time.
+* Speed up package upgrades.
+
+# Detailed design
+[design]: #detailed-design
+
+Add a new `meta.autoMerge` package attribute with type `bool` defaulting to `false`. To be documented in Nixpkgs manual.
+
+OfBorg reads `meta.autoMerge` value and sets `13.auto-merge` label to PR.
+
+GitHub Actions will trigger a merge when the following 4 conditions are met:
+1) `@r-ryantm` is PR's author.
+2) Package `meta.autoMerge` attribute is enabled.
+3) All CI checks passed.
+4) Package maintainer has approved PR.
+
+Due maintainers being able to add labels in GitHub, labels are not to be trusted and are used only to trigger GitHub Actions.
+An extra check of the 4 conditions (without using labels) is necessary in GitHub Actions before merge can happen.
+
+The appropriateness of setting `meta.autoMerge` is left up to committers.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* Reduced trustworthiness.
+* Some possible security issue.
```

Comment:
```suggestion
* Possible security issues:
  * Changing ofborg from being read-only in the nixpkgs repository (aside from setting CI status) to being a critical security related software where potential bugs in the reviewer request code could lead to rogue merges  
* packages with many maintainers and auto merge enabled could receive accidental merges from older maintainers not fully aware of the auto merge feature yet
```

Reply:
> * Changing ofborg from being read-only in the nixpkgs repository (aside from setting CI status) to being a critical security related software where potential bugs in the reviewer request code could lead to rogue merges

OfBorg is restricted to adding labels and is absent of merge process. auto-merge label is used to inform user. Labels only trigger GitHub Actions to run, rules enforcement and merge is done by some code yet to be defined. Any failure in the code used to enforce rules and merge would be a security problem, yes.

> * packages with many maintainers and auto merge enabled could receive accidental merges from older maintainers not fully aware of the auto merge feature yet

Possible solution: `meta.maintainersGrantedAutoMerge = [ maintainer1 maintainer2 ];`
Seems to me this is a case of exception, not rule. Because the maintainer would have already tested and approved code. The risk is not too high. We can always rollback some bogus change.

Reply:
> * Changing ofborg from being read-only in the nixpkgs repository (aside from setting CI status) to being a critical security related software where potential bugs in the reviewer request code could lead to rogue merges

OfBorg was restricted to adding labels and is absent of merge process. OfBorg is no longer labelling `auto-merge`, the maintainer is now responsible to add the `auto-merge` label.

The `auto-merge` labels only trigger GitHub Actions to run, rules enforcement and merge is done by some code yet to be defined in GitHub Actions. Any failure in the code used to enforce rules and merge would be a security problem, so I've added a mention for it. (`Security issues: Possible failures in the GitHub Action code used to enforce rules and merge.`)

> * packages with many maintainers and auto merge enabled could receive accidental merges from older maintainers not fully aware of the auto merge feature yet

Removed OfBorg reference. And added:
`Maintainer adds "auto-merge" label to PR triggering GitHub Actions to enforce rules and merge.`